File: finding-seams/typescript/src/badge-printing/badge-printing.ts
import { badgeSequence } from './badge-id-generation.js'
import { Badge } from './badge.js'

type PrinterDrivers = { get: (name: string) => unknown }
type PrinterConfigs = { getConfigBasedOnDimensions: (dimensions: unknown) => Promise<unknown> }
type BadgeDimensionResolvers = { resolve: (badge: Badge) => Promise<unknown> }
type PrinterFactory = {
  load: (driver: unknown, config: unknown) => Promise<{ print: (badge: Badge) => void }>
}

// Legacy-style dispenser tightly coupled to the global sequence.
export class BadgePrinting {
  #printerDrivers: PrinterDrivers | undefined
  #printerConfigs: PrinterConfigs | undefined
  #badgeDimensionResolvers: BadgeDimensionResolvers | undefined
  #printerFactory: PrinterFactory | undefined

  constructor(
    printerDrivers?: PrinterDrivers,
    printerConfigs?: PrinterConfigs,
    badgeDimensionResolvers?: BadgeDimensionResolvers,
    printerFactory?: PrinterFactory,
  ) {
    this.#printerDrivers = printerDrivers
    this.#printerConfigs = printerConfigs
    this.#badgeDimensionResolvers = badgeDimensionResolvers
    this.#printerFactory = printerFactory
  }

  createBadge(name: string, pronouns: string): Badge {
    const issuedAt = new Date()

    const id = badgeSequence.nextId()
    const formattedId = `CONF-${id}`
    return new Badge(formattedId, name, pronouns, issuedAt)
  }

  /*
  This is a very clumsy method, only to add numerous arguments for constructor, to show that constructor
  injection in this case might not be the best possible option - and should not be used for this exercise.
   */
  printBadge(badge: Badge): void {
    this.#badgeDimensionResolvers!.resolve(badge).then(dimensions => {
      this.#printerConfigs!.getConfigBasedOnDimensions(dimensions).then(config => {
        const driver = this.#printerDrivers!.get('BadgePrinting')
        this.#printerFactory!.load(driver, config).then(printer => {
          printer.print(badge)
        })
      })
    })
  }
}

File: finding-seams/python/src/badge_printing/badge_printing.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from badge_printing.badge import Badge
from badge_printing.badge_id_generation import badge_sequence


class BadgePrinting:
    """
    Legacy-style dispenser tightly coupled to the global sequence.
    """

    def __init__(
        self,
        printer_drivers: Any = None,
        printer_configs: Any = None,
        badge_dimension_resolvers: Any = None,
        printer_factory: Any = None,
    ) -> None:
        self._printer_drivers = printer_drivers
        self._printer_configs = printer_configs
        self._badge_dimension_resolvers = badge_dimension_resolvers
        self._printer_factory = printer_factory

    def create_badge(self, name: str, pronouns: str) -> Badge:
        issued_at = datetime.now()
        next_num = badge_sequence.next_id()
        formatted_id = f"CONF-{next_num}"
        return Badge(id=formatted_id, name=name, pronouns=pronouns, issued_at=issued_at)

    def print_badge(self, badge: Badge) -> None:
        """
        This is a very clumsy method, only to add numerous arguments for constructor, to show that constructor
        injection in this case might not be the best possible option - and should not be used for this exercise.
        """
        dimensions = self._badge_dimension_resolvers.resolve(badge)
        config = self._printer_configs.get_config_based_on_dimensions(dimensions)
        driver = self._printer_drivers.get("BadgePrinting")
        printer = self._printer_factory.load(driver, config)
        printer.print(badge)
